### PR TITLE
[CLI][PFCWD][Multi-ASIC] Added multi ASIC support to 'pfcwd' CLI command

### PR DESCRIPTION
**Pull request by smaheshm**
**- What I did**

Added multi ASIC support to the following CLI command:

>> show pfcwd config
>> show pfcwd stats.

Added multi ASIC support to following script:

>> pfcwd 

Commands:
  big_red_switch  Enable/disable BIG_RED_SWITCH mode
  counter_poll    Enable/disable counter polling
  interval        Set PFC watchdog counter polling interval
  show            Show PFC Watchdog information
  start           Start PFC watchdog on port(s).
  start_default   Start PFC WD by default configurations
  stop            Stop PFC watchdog on port(s)

>> 

**- How I did it**

Multi ASIC decorator is used to run on multi ASIC and single ASIC. Multi ASIC requires the CLI to be a class instance. 

Converted CLIs to classes to that multi ASIC decorator can be used.

Main logic change in 'show' command is:

* collect the stats from all ASICs.
* display the output

'show' commands have the option of passing asic instance and display filter, applicable to multi ASIC platforms.

For config changes, 

* run the config change on each asic instance on multi ASIC platforms, or the only ASIC on single ASIC platforms.

Config commands don't have the option to pass the asic instance or the display filter. By default its all instances and all ports.

**- How to verify it**

Ran the commands CLI commands on both multi ASIC and single ASIC platforms, and unit tests.

```
admin@str-n3164-acs-2:~$ pfcwd show stats --help
Usage: pfcwd show stats [OPTIONS] [QUEUES]...

  Show PFC Watchdog stats per queue

Options:
  -d, --display [all|frontend]    Show internal interfaces  [default:
                                  frontend]
  -n, --namespace [asic0|asic1|asic2|asic3|asic4|asic5]
                                  Namespace name or all
  -e, --empty
  --help                          Show this message and exit.
admin@str-n3164-acs-2:~$ pfcwd show config --help
Usage: pfcwd show config [OPTIONS] [PORTS]...

  Show PFC Watchdog configuration

Options:
  -d, --display [all|frontend]    Show internal interfaces  [default:
                                  frontend]
  -n, --namespace [asic0|asic1|asic2|asic3|asic4|asic5]
                                  Namespace name or all
  --help                          Show this message and exit.
admin@str-n3164-acs-2:~$ 

admin@str-n3164-acs-2:~$ pfcwd show config -n asic1 
Changed polling interval to 200 ms on asic1
BIG_RED_SWITCH status is disable on asic1
      PORT    ACTION    DETECTION TIME    RESTORATION TIME
----------  --------  ----------------  ------------------
Ethernet64      drop               200                 200
Ethernet68      drop               200                 200
Ethernet80      drop               200                 200
Ethernet84      drop               200                 200

admin@str-n3164-acs-2:~$ pfcwd show config -n asic1 -d all
Changed polling interval to 200 ms on asic1
BIG_RED_SWITCH status is disable on asic1
          PORT    ACTION    DETECTION TIME    RESTORATION TIME
--------------  --------  ----------------  ------------------
    Ethernet64      drop               200                 200
    Ethernet68      drop               200                 200
    .
    .
    .
Ethernet-BP112      drop               200                 200
Ethernet-BP116      drop               200                 200
Ethernet-BP120      drop               200                 200
Ethernet-BP124      drop               200                 200

admin@str-n3164-acs-2:~$ pfcwd show config 
Changed polling interval to 200 ms on asic0
BIG_RED_SWITCH status is disable on asic0
Changed polling interval to 200 ms on asic1
BIG_RED_SWITCH status is disable on asic1
Changed polling interval to 200 ms on asic2
BIG_RED_SWITCH status is disable on asic2
Changed polling interval to 200 ms on asic3
BIG_RED_SWITCH status is disable on asic3
       PORT    ACTION    DETECTION TIME    RESTORATION TIME
-----------  --------  ----------------  ------------------
Ethernet0      drop               200                 200
Ethernet4      drop               200                 200
Ethernet20      drop               200                 200
 .
 .
Ethernet64      drop               200                 200
Ethernet84      drop               200                 200
.
.
Ethernet136      drop               200                 200
Ethernet168      drop               200                 200
.
.
Ethernet180      drop               200                 200
Ethernet220      drop               200                 200
.
.
Ethernet244      drop               200                 200
Ethernet248      drop               200                 200
Ethernet252      drop               200                 200

```

pfcstat command 

```

admin@str-n3164-acs-2:~$ sudo pfcwd interval 197
admin@str-n3164-acs-2:~$ sudo pfcwd start --action drop ports all detection-time 199 --restoration-time 199
admin@str-n3164-acs-2:~$ pfcwd show 
config  stats   
admin@str-n3164-acs-2:~$ pfcwd show config
Changed polling interval to 197 ms on asic0
BIG_RED_SWITCH status is enable on asic0
Changed polling interval to 197 ms on asic1
BIG_RED_SWITCH status is enable on asic1
Changed polling interval to 197 ms on asic2
BIG_RED_SWITCH status is enable on asic2
Changed polling interval to 197 ms on asic3
BIG_RED_SWITCH status is enable on asic3
       PORT    ACTION    DETECTION TIME    RESTORATION TIME
-----------  --------  ----------------  ------------------
  Ethernet0      drop               199                 199
  Ethernet4      drop               199                 199
  Ethernet8      drop               199                 199
 Ethernet12      drop               199                 199
 Ethernet16      drop               199                 199
 Ethernet20      drop               199                 199
 Ethernet24      drop               199                 199
 Ethernet28      drop               199                 199
 Ethernet32      drop               199                 199
 .
 .
 .
 .
Ethernet228      drop               199                 199
Ethernet232      drop               199                 199
Ethernet236      drop               199                 199
Ethernet240      drop               199                 199
Ethernet244      drop               199                 199
Ethernet248      drop               199                 199
Ethernet252      drop               199                 199
admin@str-n3164-acs-2:~$ 


```

### CODE REVIEW

**Comment by smaheshm** [comment]
Working on unit test code, raised the PR so review can start.

File: show/main.py
Comment (smaheshm) [inline show/main.py:416]: Will fix the args.

File: show/main.py
Comment (smaheshm) [inline show/main.py:420]: will add check for namespace 'None'

File: show/main.py
Comment (smaheshm) [inline show/main.py:430]: will add check for namespace 'None'

**Comment by smaheshm** [comment]
Unit tests are written only for 'show' commands. Unit tests for config commands is pending on mock config set implementation for multi ASIC environment.

File: pfcwd/main.py
Comment (smaheshm) [inline pfcwd/main.py:215]: not required, will remove this in next iteration.

File: pfcwd/main.py
Comment (smaheshm) [inline pfcwd/main.py:312]: not requird.

File: pfcwd/main.py
Comment (jleveque) [inline pfcwd/main.py:13]: Please reorganize imports per [PEP8 standards](https://www.python.org/dev/peps/pep-0008/#imports):

> Imports should be grouped in the following order:

> Standard library imports.
> Related third party imports.
> Local application/library specific imports.
> You should put a blank line between each group of imports.

And group alphabetically in each group. This should read:

```python
import os
import sys

import click
import swsssdk
from natsort import natsorted
from sonic_py_common.multi_asic import get_external_ports
from tabulate import tabulate
from utilities_common import multi_asic as multi_asic_util
from utilities_common import constants
```

File: pfcwd/main.py
Comment (jleveque) [inline pfcwd/main.py:68]: Remove "end"

File: pfcwd/main.py
Comment (jleveque) [inline pfcwd/main.py:386]: What is the benefit in encapsulating these groups inside classes?

Comment (smaheshm) [reply pfcwd/main.py:386]: Main benefit is adding sub commands becomes easier and are logically grouped. Besides, multi ASIC decorator requires a class instance to operate upon. I also thought of adding a base class like 'SonicCli' but then decided against it as the only user is 'multi ASIC' decorator which does not apply to all the CLIs.

File: pfcwd/main.py
Comment (smaheshm) [inline pfcwd/main.py:473]: will  remove

**Comment by smaheshm** [comment]
will wait for reviews from @arlakshm and @neethajohn and make suggested changes.

File: pfcwd/main.py
Comment (neethajohn) [inline pfcwd/main.py:252]: since we are not specifying the namespace option for config commands, aren't we updating the global db? How does 'show config' for a namespace work with this approach?

File: tests/pfcwd_masic_test.py
Comment (neethajohn) [inline tests/pfcwd_masic_test.py:133]: Can you include cases to cover all the options for 'show stats' and 'show config'. display for a namespace, display when a set of  ports/queues are included?

File: pfcwd/main.py
Comment (smaheshm) [reply pfcwd/main.py:252]: for all the 'config' commands, the command is for all namespaces, and the DB within the namespace is updated. 

By default, 'show config'  will fetch from namespaces that have external ports.

This logic emulates a 'single' switch interface.

File: pfcwd/main.py
Comment (smaheshm) [reply pfcwd/main.py:408]: when the 'ports' liist is passed there's no need to pass the 'display all' or 'namespace' list. The command will match ports from each namespace, and fetch config if port is present in that namespace.

File: tests/pfcstat_test.py
Comment (smaheshm) [reply tests/pfcstat_test.py:44]: I had to add some counters and some more info in coutners db to make the PFCWD utests work, which caused the output to be different. So had to make changes here as well.

File: tests/pfcwd_masic_test.py
Comment (smaheshm) [reply tests/pfcwd_masic_test.py:154]: The "display" option will not work. Requires a bit of explanation but it has to do with 'multi-asic-click-options' which populates the parameters before the environment variables are set un 'setup()' method. Due to this the multi asic namespace options are not available when the pytest framework loads all the .py files. Tried several options including 'reload', but it didn't work. We are dicussing internally what other options are available including changing implementation of 'multi-asic-click-option'.

**Comment by smaheshm** [comment]
retest this please

**Comment by smaheshm** [comment]
Just realized pytest needs file names to start or end with “test”. Will rename file.

File: pfcwd/main.py
Comment (neethajohn) [inline pfcwd/main.py:173]: Do we need a new var? Wouldn't checking 'table' empty suffice?

Comment (smaheshm) [reply pfcwd/main.py:173]: table may not be empty if multiple ports on different ASICs are specified.

**Comment by smaheshm** [comment]
This passes @neethajohn recently added unit tests on single ASIC platform.

**Review by jleveque — APPROVED** [review]
Looks good from my perspective. Please wait for other reviewers.

File: tests/pfcwd_masic_test.py
Comment (neethajohn) [reply tests/pfcwd_masic_test.py]: Would these changes come in a following PR?

Comment (smaheshm) [reply tests/pfcwd_masic_test.py]: Yes.

Comment (smaheshm) [reply tests/pfcwd_masic_test.py]: #1093

**Review by neethajohn — APPROVED** [review]

**Comment by abdosi** [comment]
@smaheshm Create PR for 201911

**Comment by smaheshm** [comment]
> @smaheshm Create PR for 201911

#1102

**Pull request merged**
